app/controllers/katello/api/v2/content_view_components_controller.rb

```diff
@@ -0,0 +1,100 @@
+    param :composite_content_view_id, :identifier, :desc => N_("composite content view identifier"), :required => true
+    param :components, Array, :desc => N_("Array of components to add"), :required => true do
+      param :content_view_version_id, :identifier, :desc => N_("identifier of the version of the component content view")
+      param :content_view_id, :identifier,
+            :desc => N_("content view identifier of the component who's latest version is desired")
+      param :latest, :bool, :desc => N_("true if the latest version of the components content view is desired")
+    end
+    def add_components
+      components = params.require(:components).map do |component|
+        options = {}
+        options[:latest] = ::Foreman::Cast.to_bool(component[:latest]) if component.key?(:latest)
+        options.merge(component.slice(:content_view_version_id, :content_view_id))
+      end
+      @view.add_components(components)
+      @view.save!
+      respond_for_index(:collection => index_response, :template => "index")
+    end
+
+    api :PUT, "/content_views/:composite_content_view_id/content_view_components/remove",
+                                        N_("Remove components from the content view")
+    param :composite_content_view_id, :identifier, :desc => N_("composite content view identifier"), :required => true
+    param :component_ids, Array, :desc => N_("Array of component ids to remove"), :required => true
+    def remove_components
+      @view.remove_components(params.require(:component_ids))
+      @view.save!
+      respond_for_index(:collection => index_response, :template => "index")
+    end
+
+    api :GET, "/content_views/:composite_content_view_id/content_view_components/:id", N_("Show a content view component")
+    param :composite_content_view_id, :number, :desc => N_("composite content view numeric identifier"), :required => true
+    param :id, :identifier, :desc => N_("component ID"), :required => true
+    def show
+      respond :resource => @component
+    end
+
+    api :PUT, "/content_views/:composite_content_view_id/content_view_components/:id",
+        N_("Update a component associated with the content view")
+    param :composite_content_view_id, :identifier, :desc => N_("composite content view identifier"), :required => true
+    param :id, :identifier, :desc => N_("component ID"), :required => true
+    param :content_view_version_id, :identifier, :desc => N_("identifier of the version of the component content view")
+    param :latest, :bool, :desc => N_("true if the latest version of the components content view is desired")
+    def update
+      cvv_id = component_params[:content_view_version_id]
+      if component_params.key?(:latest) && component_params.key?(:content_view_version_id)
+        latest = ::Foreman::Cast.to_bool(component_params[:latest])
+        if latest && cvv_id.present?
+          fail HttpErrors::UnprocessableEntity,
```

Comment:
@bbuckingham so I debated about this. 
3 cases to think of where we are updating an existing component.

1. params = {latest: true}
1. params = {latest: true, content_view_version_id => same_as_old_version_id }
1. params = {latest: true, content_view_version_id => different_version_id }

Now if I were to move this controller logic to model, I 'd have to add a before_validate that says
```
if latest was changed then clear out the content view version
```

but now if the request both latest=true and  had content_view_version_id we'd need to fail the save in that case. A model will not know that  a content_view_version_id got passed in the request . Since its in the request itself I believe its best suited in the controller.

Model level validation just makes sure that the params are set in a valid state before save.